File: src/sources/fastmail.ts
#!/usr/bin/env bun
import {
  sourceMain,
  type IntakeItem,
  type PollRequest,
  type PollResponse,
} from "../protocol.ts"

interface JmapSession {
  apiUrl: string
  primaryAccounts: Record<string, string>
}

interface Checkpoint {
  queryState: string
  mailboxId: string
  sentMailboxId?: string | null
}

interface Email {
  id: string
  threadId: string
  subject?: string
  from?: Address[]
  to?: Address[]
  cc?: Address[]
  bcc?: Address[]
  receivedAt: string
  sentAt?: string
  messageId?: string[]
  mailboxIds?: Record<string, boolean>
  bodyValues?: Record<string, { value: string; isTruncated?: boolean }>
  textBody?: Array<{ partId?: string; type?: string }>
  htmlBody?: Array<{ partId?: string; type?: string }>
  bodyStructure?: BodyPart
}

interface HeaderRule {
  property: string
  values: Set<string>
}

interface EmailFilters {
  includeHeaders: HeaderRule[]
  excludeHeaders: HeaderRule[]
  includeMessageIdContains: string[]
}

interface Address {
  name?: string
  email: string
}

interface BodyPart {
  partId?: string
  blobId?: string
  name?: string
  type?: string
  size?: number
  disposition?: string
  cid?: string
  subParts?: BodyPart[]
}

const EMAIL_ACCOUNT_CAPABILITY = "urn:ietf:params:jmap:mail"
const BODY_LIMIT = 64 * 1024
const THREAD_MESSAGE_LIMIT = 100
const ATTACHMENT_LIMIT = 100

export async function pollFastmail(
  request: PollRequest,
  fetcher: typeof fetch = fetch,
): Promise<PollResponse> {
  const token = process.env.FASTMAIL_API_TOKEN
  if (!token) throw new Error("FASTMAIL_API_TOKEN is required")
  const sessionUrl =
    stringOption(request, "session_url") ??
    "https://api.fastmail.com/jmap/session"
  const sessionResponse = await fetcher(sessionUrl, {
    headers: { Authorization: `Bearer ${token}` },
  })
  if (!sessionResponse.ok)
    throw new Error(
      `Fastmail JMAP session request failed with ${sessionResponse.status}`,
    )
  const session = (await sessionResponse.json()) as JmapSession
  const accountId =
    stringOption(request, "account_id") ??
    session.primaryAccounts[EMAIL_ACCOUNT_CAPABILITY]
  if (!accountId) throw new Error("Fastmail JMAP session has no mail account")
  const checkpoint = request.checkpoint
    ? parseCheckpoint(request.checkpoint)
    : undefined
  let mailboxId = stringOption(request, "mailbox_id") ?? checkpoint?.mailboxId
  let sentMailboxId = checkpoint?.sentMailboxId
  if (!mailboxId || sentMailboxId === undefined) {
    const mailboxes = await findMailboxes(
      session.apiUrl,
      accountId,
      token,
      fetcher,
    )
    mailboxId ??= mailboxes.inboxId
    sentMailboxId = mailboxes.sentId
  }
  if (!mailboxId) throw new Error("Fastmail account has no inbox mailbox")
  const bootstrapLimit = Math.min(
    integerOption(request, "bootstrap_limit") ?? 0,
    request.itemLimit,
  )
  const filters = emailFilters(request)

  if (!request.checkpoint) {
    const baseline = await queryMailbox(
      session.apiUrl,
      accountId,
      token,
      mailboxId,
      bootstrapLimit,
      fetcher,
    )
    const messages =
      baseline.ids.length > 0
        ? await getEmails(
            session.apiUrl,
            accountId,
            token,
            baseline.ids,
            filters,
            fetcher,
          )
        : []
    return {
      protocolVersion: 1,
      checkpoint: { queryState: baseline.queryState, mailboxId, sentMailboxId },
      items: await normalizeMessages(
        session.apiUrl,
        accountId,
        token,
        messages,
        sentMailboxId,
        filters,
        fetcher,
      ),
    }
  }

  if (!checkpoint || checkpoint.mailboxId !== mailboxId) {
    const baseline = await queryMailbox(
      session.apiUrl,
      accountId,
      token,
      mailboxId,
      0,
      fetcher,
    )
    return {
      protocolVersion: 1,
      checkpoint: { queryState: baseline.queryState, mailboxId, sentMailboxId },
      items: [],
    }
  }

  const changes = await jmapCall<any>(
    session.apiUrl,
    token,
    [
      "Email/queryChanges",
      {
        accountId,
        filter: { inMailbox: mailboxId },
        sort: [{ property: "receivedAt", isAscending: false }],
        sinceQueryState: checkpoint.queryState,
        maxChanges: request.itemLimit,
      },
      "changes",
    ],
    fetcher,
  )
  if (typeof changes.newQueryState !== "string")
    throw new Error("Fastmail query changes response has no query state")
  const ids = (changes.added ?? [])
    .map((addition: { id?: string }) => addition.id)
    .filter((id: unknown): id is string => typeof id === "string")
    .slice(0, request.itemLimit)
  const messages =
    ids.length > 0
      ? await getEmails(session.apiUrl, accountId, token, ids, filters, fetcher)
      : []

  return {
    protocolVersion: 1,
    checkpoint: { queryState: changes.newQueryState, mailboxId, sentMailboxId },
    items: await normalizeMessages(
      session.apiUrl,
      accountId,
      token,
      messages,
      sentMailboxId,
      filters,
      fetcher,
    ),
  }
}

async function queryMailbox(
  apiUrl: string,
  accountId: string,
  token: string,
  mailboxId: string,
  limit: number,
  fetcher: typeof fetch,
): Promise<{ queryState: string; ids: string[] }> {
  const result = await jmapCall<any>(
    apiUrl,
    token,
    [
      "Email/query",
      {
        accountId,
        filter: { inMailbox: mailboxId },
        sort: [{ property: "receivedAt", isAscending: false }],
        limit,
      },
      "query",
    ],
    fetcher,
  )
  if (typeof result.queryState !== "string")
    throw new Error("Fastmail mailbox query response has no query state")
  return { queryState: result.queryState, ids: result.ids ?? [] }
}

async function normalizeMessages(
  apiUrl: string,
  accountId: string,
  token: string,
  messages: Email[],
  sentMailboxId: string | null,
  filters: EmailFilters,
  fetcher: typeof fetch,
): Promise<IntakeItem[]> {
  const threadCache = new Map<string, Email[]>()
  const items: IntakeItem[] = []
  const ordered = [...messages].sort((left, right) =>
    messageTimestamp(left).localeCompare(messageTimestamp(right)),
  )
  for (const email of ordered) {
    if (!isAllowed(email, filters)) continue
    let thread = threadCache.get(email.threadId)
    if (!thread) {
      thread = await getThread(
        apiUrl,
        accountId,
        token,
        email.threadId,
        filters,
        fetcher,
      )
      threadCache.set(email.threadId, thread)
    }
    const latest = thread.at(-1)
    if (latest && sentMailboxId && latest.mailboxIds?.[sentMailboxId]) continue
    items.push(normalizeEmail(accountId, email, thread))
  }
  return items
}

async function findMailboxes(
  apiUrl: string,
  accountId: string,
  token: string,
  fetcher: typeof fetch,
): Promise<{ inboxId: string | undefined; sentId: string | null }> {
  const result = await jmapCall<any>(
    apiUrl,
    token,
    ["Mailbox/get", { accountId, properties: ["id", "role"] }, "mailboxes"],
    fetcher,
  )
  const mailboxes = (result.list ?? []) as Array<{
    id?: string
    role?: string | null
  }>
  return {
    inboxId: mailboxes.find((mailbox) => mailbox.role === "inbox")?.id,
    sentId: mailboxes.find((mailbox) => mailbox.role === "sent")?.id ?? null,
  }
}

async function getEmails(
  apiUrl: string,
  accountId: string,
  token: string,
  ids: string[],
  filters: EmailFilters,
  fetcher: typeof fetch,
): Promise<Email[]> {
  const result = await jmapCall<any>(
    apiUrl,
    token,
    [
      "Email/get",
      {
        accountId,
        ids,
        properties: [
          "id",
          "threadId",
          "subject",
          "from",
          "to",
          "cc",
          "bcc",
          "receivedAt",
          "sentAt",
          "messageId",
          "mailboxIds",
          "textBody",
          "htmlBody",
          "bodyValues",
          "bodyStructure",
          ...new Set(
            [...filters.includeHeaders, ...filters.excludeHeaders].map(
              (rule) => rule.property,
            ),
          ),
        ],
        fetchTextBodyValues: true,
        fetchHTMLBodyValues: true,
        maxBodyValueBytes: BODY_LIMIT,
      },
      "emails",
    ],
    fetcher,
  )
  return result.list ?? []
}

async function getThread(
  apiUrl: string,
  accountId: string,
  token: string,
  threadId: string,
  filters: EmailFilters,
  fetcher: typeof fetch,
): Promise<Email[]> {
  const thread = await jmapCall<any>(
    apiUrl,
    token,
    ["Thread/get", { accountId, ids: [threadId] }, "thread"],
    fetcher,
  )
  const ids = (thread.list?.[0]?.emailIds ?? []).slice(-THREAD_MESSAGE_LIMIT)
  const emails = await getEmails(
    apiUrl,
    accountId,
    token,
    ids,
    filters,
    fetcher,
  )
  return emails
    .filter((email) => isAllowed(email, filters))
    .sort((left, right) =>
      messageTimestamp(left).localeCompare(messageTimestamp(right)),
    )
}

async function jmapCall<T>(
  apiUrl: string,
  token: string,
  methodCall: [string, Record<string, unknown>, string],
  fetcher: typeof fetch,
): Promise<T> {
  const response = await fetcher(apiUrl, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      using: [EMAIL_ACCOUNT_CAPABILITY, "urn:ietf:params:jmap:core"],
      methodCalls: [methodCall],
    }),
  })
  if (!response.ok)
    throw new Error(`Fastmail JMAP request failed with ${response.status}`)
  const json = (await response.json()) as {
    methodResponses?: Array<
      [string, T & { type?: string; description?: string }]
    >
  }
  const result = json.methodResponses?.[0]
  if (!result) throw new Error("Fastmail JMAP response has no method result")
  if (result[0] === "error")
    throw new Error(`Fastmail JMAP error: ${result[1].type ?? "unknown"}`)
  return result[1]
}

function messageTimestamp(email: Email): string {
  return email.sentAt ?? email.receivedAt
}

function normalizeEmail(
  accountId: string,
  email: Email,
  thread: Email[],
): IntakeItem {
  const threadBody = thread
    .map((message) => {
      const sender = formatAddresses(message.from)
      const recipients = formatAddresses([
        ...(message.to ?? []),
        ...(message.cc ?? []),
        ...(message.bcc ?? []),
      ])
      return `From: ${sender}\nTo: ${recipients}\nDate: ${message.receivedAt}\nSubject: ${message.subject ?? "(no subject)"}\n\n${bodyText(message)}`
    })
    .join("\n\n---\n\n")
    .slice(0, BODY_LIMIT * 4)
  const attachments = thread
    .flatMap((message) => attachmentMetadata(message.bodyStructure))
    .slice(0, ATTACHMENT_LIMIT)
  return {
    entityId: `fastmail:${accountId}:thread:${email.threadId}`,
    revisionId: email.id,
    kind: "email",
    title: email.subject ?? "(no subject)",
    body: threadBody,
    occurredAt: email.receivedAt,
    metadata: {
      messageId: email.id,
      threadId: email.threadId,
      from: email.from ?? [],
      to: email.to ?? [],
      cc: email.cc ?? [],
      bcc: email.bcc ?? [],
      attachments,
      threadMessageCount: thread.length,
    },
  }
}

function bodyText(email: Email): string {
  const parts = email.textBody?.length ? email.textBody : (email.htmlBody ?? [])
  const values = parts.map((part) =>
    part.partId ? (email.bodyValues?.[part.partId]?.value ?? "") : "",
  )
  return values.join("\n").slice(0, BODY_LIMIT)
}

function attachmentMetadata(
  root?: BodyPart,
): Array<Record<string, string | number | null>> {
  if (!root) return []
  const result: Array<Record<string, string | number | null>> = []
  const pending = [root]
  while (pending.length > 0 && result.length < ATTACHMENT_LIMIT) {
    const part = pending.pop()
    if (!part) continue
    pending.push(...(part.subParts ?? []))
    if (!part.blobId || (!part.name && part.disposition !== "attachment"))
      continue
    result.push({
      name: part.name ?? null,
      type: part.type ?? "application/octet-stream",
      size: part.size ?? 0,
      disposition: part.disposition ?? null,
      cid: part.cid ?? null,
    })
  }
  return result
}

function formatAddresses(addresses?: Address[]): string {
  return (addresses ?? [])
    .map((address) =>
      address.name ? `${address.name} <${address.email}>` : address.email,
    )
    .join(", ")
}

function parseCheckpoint(value: unknown): Checkpoint | undefined {
  if (!value || typeof value !== "object")
    throw new Error("Fastmail checkpoint is invalid")
  const checkpoint = value as Partial<Checkpoint>
  if (
    typeof checkpoint.queryState !== "string" ||
    typeof checkpoint.mailboxId !== "string" ||
    (checkpoint.sentMailboxId !== undefined &&
      checkpoint.sentMailboxId !== null &&
      typeof checkpoint.sentMailboxId !== "string")
  ) {
    return undefined
  }
  return checkpoint as Checkpoint
}

function emailFilters(request: PollRequest): EmailFilters {
  return {
    includeHeaders: headerRules(request, "include_headers"),
    excludeHeaders: headerRules(request, "exclude_headers"),
    includeMessageIdContains: stringListOption(
      request,
      "include_message_id_contains",
    ).map((value) => value.toLowerCase()),
  }
}

function headerRules(request: PollRequest, optionName: string): HeaderRule[] {
  const option = request.options[optionName]
  if (option === undefined) return []
  if (!option || typeof option !== "object" || Array.isArray(option))
    throw new Error(`Fastmail ${optionName} must map header names to values`)

  return Object.entries(option).map(([name, values]) => {
    if (!/^[A-Za-z0-9-]+$/.test(name))
      throw new Error(`Fastmail ${optionName} header name is invalid: ${name}`)
    if (
      !Array.isArray(values) ||
      values.length === 0 ||
      values.some((value) => typeof value !== "string" || !value.trim())
    )
      throw new Error(
        `Fastmail ${optionName} header ${name} must have a non-empty value list`,
      )
    return {
      property: `header:${name}:asText`,
      values: new Set(
        values.map((value) => (value as string).trim().toLowerCase()),
      ),
    }
  })
}

function isAllowed(email: Email, filters: EmailFilters): boolean {
  const includedByMessageId =
    filters.includeMessageIdContains.length === 0 ||
    (email.messageId ?? []).some((messageId) =>
      filters.includeMessageIdContains.some((value) =>
        messageId.toLowerCase().includes(value),
      ),
    )
  return (
    includedByMessageId &&
    filters.includeHeaders.every((rule) => matchesHeaderRule(email, rule)) &&
    !filters.excludeHeaders.some((rule) => matchesHeaderRule(email, rule))
  )
}

function matchesHeaderRule(email: Email, rule: HeaderRule): boolean {
  const value = Reflect.get(email, rule.property)
  return (
    typeof value === "string" && rule.values.has(value.trim().toLowerCase())
  )
}

function stringListOption(request: PollRequest, name: string): string[] {
  const value = request.options[name]
  if (value === undefined) return []
  if (
    !Array.isArray(value) ||
    value.length === 0 ||
    value.some((item) => typeof item !== "string" || !item.trim())
  )
    throw new Error(`Fastmail ${name} must be a non-empty string list`)
  return value.map((item) => (item as string).trim())
}

function stringOption(request: PollRequest, name: string): string | undefined {
  const value = request.options[name]
  return typeof value === "string" ? value : undefined
}

function integerOption(request: PollRequest, name: string): number | undefined {
  const value = request.options[name]
  return typeof value === "number" && Number.isSafeInteger(value) && value >= 0
    ? value
    : undefined
}

if (import.meta.main) sourceMain(pollFastmail)

File: tests/fastmail.rs
mod support;

use std::process::Stdio;

use intake::protocol::{PollRequest, PollResponse};
use intake::sources::{fastmail::poll_fastmail, http_client};
use serde_json::{Map, Value, json};
use support::FixtureServer;
use tokio::io::AsyncWriteExt;
use tokio::process::Command;

fn request(base_url: &str, checkpoint: Value) -> PollRequest {
    PollRequest {
        protocol_version: 1,
        source: "fastmail".into(),
        checkpoint,
        now: "2026-08-03T10:10:00.000Z".into(),
        item_limit: 10,
        options: Map::from_iter([
            ("mailbox_id".into(), json!("inbox")),
            ("session_url".into(), json!(format!("{base_url}/session"))),
        ]),
    }
}

fn session(base_url: &str) -> Value {
    json!({
        "apiUrl": format!("{base_url}/jmap"),
        "primaryAccounts": { "urn:ietf:params:jmap:mail": "account-1" },
    })
}

fn jmap(method: &str, value: Value, call_id: &str) -> Value {
    json!({ "methodResponses": [[method, value, call_id]] })
}

fn email(id: &str, received_at: &str, value: &str) -> Value {
    json!({
        "id": id,
        "threadId": "thread-1",
        "subject": "Question",
        "from": [{ "name": "Sender", "email": "sender@example.test" }],
        "to": [{ "email": "recipient@example.test" }],
        "receivedAt": received_at,
        "mailboxIds": { "inbox": true },
        "textBody": [{ "partId": "body", "type": "text/plain" }],
        "bodyValues": { "body": { "value": value } },
        "bodyStructure": {
            "subParts": [{
                "blobId": "blob-secret",
                "name": "report.pdf",
                "type": "application/pdf",
                "size": 42,
                "disposition": "attachment",
            }],
        },
    })
}

#[tokio::test]
async fn establishes_a_mailbox_query_baseline_without_historical_events() {
    let server = FixtureServer::start(|base| {
        vec![
            session(base),
            jmap(
                "Mailbox/get",
                json!({
                    "list": [
                        { "id": "inbox", "role": "inbox" },
                        { "id": "sent", "role": "sent" },
                    ],
                }),
                "mailboxes",
            ),
            jmap(
                "Email/query",
                json!({ "queryState": "query-state-1", "ids": [] }),
                "query",
            ),
        ]
    })
    .await;
    let result = poll_fastmail(
        request(&server.base_url, Value::Null),
        &http_client().unwrap(),
        "source-only-token",
    )
    .await
    .unwrap();
    assert!(result.items.is_empty());
    assert_eq!(
        result.checkpoint,
        json!({
            "queryState": "query-state-1",
            "mailboxId": "inbox",
            "sentMailboxId": "sent",
        })
    );
    let calls = server.finish().await;
    assert_eq!(calls.len(), 3);
    assert_eq!(calls[0].method, "GET");
    assert_eq!(calls[0].target, "/session");
    assert_eq!(calls[1].method, "POST");
    assert_eq!(calls[1].target, "/jmap");
    assert!(
        calls
            .iter()
            .all(|call| call.headers.contains("Bearer source-only-token"))
    );
    assert_eq!(calls[1].body["methodCalls"][0][0], "Mailbox/get");
}

#[tokio::test]
async fn emits_stable_events_with_bounded_threads_and_attachment_metadata() {
    let message_1 = {
        let mut value = email("message-1", "2026-08-03T10:00:00.000Z", "Initial request");
        value["sentAt"] = json!("2026-08-03T10:00:00.000Z");
        value["mailboxIds"] = json!({ "sent": true });
        value
    };
    let message_2 = email("message-2", "2026-08-03T10:05:00.000Z", "Follow up");
    let server = FixtureServer::start(|base| {
        vec![
            session(base),
            jmap(
                "Email/queryChanges",
                json!({
                    "added": [{ "id": "message-2", "index": 0 }],
                    "removed": [],
                    "newQueryState": "query-state-2",
                    "hasMoreChanges": false,
                }),
                "changes",
            ),
            jmap(
                "Email/get",
                json!({ "list": [message_2.clone()] }),
                "emails",
            ),
            jmap(
                "Thread/get",
                json!({ "list": [{ "id": "thread-1", "emailIds": ["message-1", "message-2"] }] }),
                "thread",
            ),
            jmap(
                "Email/get",
                json!({ "list": [message_1, message_2] }),
                "emails",
            ),
        ]
    })
    .await;
    let result = poll_fastmail(
        request(
            &server.base_url,
            json!({ "queryState": "query-state-1", "mailboxId": "inbox", "sentMailboxId": "sent" }),
        ),
        &http_client().unwrap(),
        "source-only-token",
    )
    .await
    .unwrap();
    assert_eq!(result.items.len(), 1);
    let item = &result.items[0];
    assert_eq!(item.entity_id, "fastmail:account-1:thread:thread-1");
    assert_eq!(item.revision_id, "message-2");
    assert!(item.body.contains("Initial request"));
    assert!(item.body.contains("Follow up"));
    assert_eq!(item.metadata["attachments"].as_array().unwrap().len(), 2);
    assert!(
        !serde_json::to_string(&item.metadata)
            .unwrap()
            .contains("blob-secret")
    );
    let calls = server.finish().await;
    assert_eq!(calls[1].body["methodCalls"][0][1]["maxChanges"], 10);
    assert_eq!(
        calls[1].body["methodCalls"][0][1]["filter"],
        json!({ "inMailbox": "inbox" })
    );
}

#[tokio::test]
async fn excludes_messages_by_header_before_thread_assembly() {
    let mut push = email(
        "push-message",
        "2026-08-03T10:05:00.000Z",
        "Pushed one commit",
    );
    push["header:X-GitHub-Reason:asText"] = json!("push");
    let server = FixtureServer::start(|base| {
        vec![
            session(base),
            jmap(
                "Email/queryChanges",
                json!({ "added": [{ "id": "push-message", "index": 0 }], "removed": [], "newQueryState": "query-state-2", "hasMoreChanges": false }),
                "changes",
            ),
            jmap("Email/get", json!({ "list": [push] }), "emails"),
        ]
    })
    .await;
    let mut request = request(
        &server.base_url,
        json!({ "queryState": "query-state-1", "mailboxId": "inbox", "sentMailboxId": "sent" }),
    );
    request.options.insert(
        "exclude_headers".into(),
        json!({ "X-GitHub-Reason": ["push"] }),
    );
    let result = poll_fastmail(request, &http_client().unwrap(), "source-only-token")
        .await
        .unwrap();
    assert!(result.items.is_empty());
    let calls = server.finish().await;
    assert!(
        calls[2].body["methodCalls"][0][1]["properties"]
            .as_array()
            .unwrap()
            .contains(&json!("header:X-GitHub-Reason:asText"))
    );
}

#[tokio::test]
async fn allowlists_headers_and_github_resource_types() {
    let mut comment = email(
        "comment-message",
        "2026-08-03T10:00:00.000Z",
        "Useful comment",
    );
    comment["header:X-GitHub-Reason:asText"] = json!("comment");
    comment["messageId"] = json!(["raine/example/issues/1/123@github.com"]);
    let mut subscribed = email(
        "subscribed-message",
        "2026-08-03T10:01:00.000Z",
        "Useful pull request",
    );
    subscribed["header:X-GitHub-Reason:asText"] = json!("subscribed");
    subscribed["messageId"] = json!(["raine/example/pull/2@github.com"]);
    let mut mention = email(
        "mention-message",
        "2026-08-03T10:02:00.000Z",
        "Noisy mention",
    );
    mention["header:X-GitHub-Reason:asText"] = json!("mention");
    mention["messageId"] = json!(["raine/example/issues/3@github.com"]);
    let mut release = email(
        "release-message",
        "2026-08-03T10:03:00.000Z",
        "Noisy release",
    );
    release["header:X-GitHub-Reason:asText"] = json!("subscribed");
    release["messageId"] = json!(["raine/example/releases/123@github.com"]);
    let messages = vec![comment, subscribed, mention, release];
    let ids: Vec<Value> = messages
        .iter()
        .map(|message| message["id"].clone())
        .collect();
    let server_messages = messages.clone();
    let server = FixtureServer::start(move |base| {
        vec![
            session(base),
            jmap(
                "Email/queryChanges",
                json!({ "added": ids.iter().enumerate().map(|(index, id)| json!({ "id": id, "index": index })).collect::<Vec<_>>(), "removed": [], "newQueryState": "query-state-2", "hasMoreChanges": false }),
                "changes",
            ),
            jmap("Email/get", json!({ "list": server_messages.clone() }), "emails"),
            jmap(
                "Thread/get",
                json!({ "list": [{ "id": "thread-1", "emailIds": ids }] }),
                "thread",
            ),
            jmap("Email/get", json!({ "list": server_messages }), "emails"),
        ]
    })
    .await;
    let mut request = request(
        &server.base_url,
        json!({ "queryState": "query-state-1", "mailboxId": "inbox", "sentMailboxId": "sent" }),
    );
    request.options.insert(
        "include_headers".into(),
        json!({ "X-GitHub-Reason": ["comment", "subscribed"] }),
    );
    request.options.insert(
        "include_message_id_contains".into(),
        json!(["/issues/", "/pull/"]),
    );
    let result = poll_fastmail(request, &http_client().unwrap(), "source-only-token")
        .await
        .unwrap();
    assert_eq!(
        result
            .items
            .iter()
            .map(|item| item.revision_id.as_str())
            .collect::<Vec<_>>(),
        ["comment-message", "subscribed-message"]
    );
    assert!(!result.items[0].body.contains("Noisy mention"));
    assert!(!result.items[0].body.contains("Noisy release"));
    server.finish().await;
}

#[tokio::test]
async fn omits_excluded_messages_from_later_thread_context() {
    let mut push = email(
        "push-message",
        "2026-08-03T10:00:00.000Z",
        "Pushed one commit",
    );
    push["header:X-GitHub-Reason:asText"] = json!("push");
    let mut comment = email(
        "comment-message",
        "2026-08-03T10:05:00.000Z",
        "Useful review comment",
    );
    comment["header:X-GitHub-Reason:asText"] = json!("subscribed");
    let server = FixtureServer::start(|base| {
        vec![
            session(base),
            jmap("Email/queryChanges", json!({ "added": [{ "id": "comment-message", "index": 0 }], "removed": [], "newQueryState": "query-state-2", "hasMoreChanges": false }), "changes"),
            jmap("Email/get", json!({ "list": [comment.clone()] }), "emails"),
            jmap("Thread/get", json!({ "list": [{ "id": "thread-1", "emailIds": ["push-message", "comment-message"] }] }), "thread"),
            jmap("Email/get", json!({ "list": [push, comment] }), "emails"),
        ]
    }).await;
    let mut request = request(
        &server.base_url,
        json!({ "queryState": "query-state-1", "mailboxId": "inbox", "sentMailboxId": "sent" }),
    );
    request.options.insert(
        "exclude_headers".into(),
        json!({ "X-GitHub-Reason": ["push"] }),
    );
    let result = poll_fastmail(request, &http_client().unwrap(), "source-only-token")
        .await
        .unwrap();
    assert_eq!(result.items.len(), 1);
    assert!(result.items[0].body.contains("Useful review comment"));
    assert!(!result.items[0].body.contains("Pushed one commit"));
    server.finish().await;
}

#[tokio::test]
async fn suppresses_a_thread_when_its_newest_message_is_sent_mail() {
    let incoming = email("message-1", "2026-08-03T10:00:00.000Z", "Initial request");
    let mut reply = email("message-2", "2026-08-03T10:05:00.000Z", "Sent reply");
    reply["sentAt"] = json!("2026-08-03T10:06:00.000Z");
    reply["mailboxIds"] = json!({ "sent": true });
    let server = FixtureServer::start(|base| vec![
        session(base),
        jmap("Mailbox/get", json!({ "list": [{ "id": "inbox", "role": "inbox" }, { "id": "sent", "role": "sent" }] }), "mailboxes"),
        jmap("Email/queryChanges", json!({ "added": [{ "id": "message-1", "index": 0 }], "removed": [], "newQueryState": "query-state-2", "hasMoreChanges": false }), "changes"),
        jmap("Email/get", json!({ "list": [incoming.clone()] }), "emails"),
        jmap("Thread/get", json!({ "list": [{ "id": "thread-1", "emailIds": ["message-1", "message-2"] }] }), "thread"),
        jmap("Email/get", json!({ "list": [incoming, reply] }), "emails"),
    ]).await;
    let result = poll_fastmail(
        request(
            &server.base_url,
            json!({ "queryState": "query-state-1", "mailboxId": "inbox" }),
        ),
        &http_client().unwrap(),
        "source-only-token",
    )
    .await
    .unwrap();
    assert!(result.items.is_empty());
    assert_eq!(result.checkpoint["sentMailboxId"], "sent");
    server.finish().await;
}

#[tokio::test]
async fn advances_through_removals_without_emitting_items() {
    let server = FixtureServer::start(|base| vec![
        session(base),
        jmap("Email/queryChanges", json!({ "added": [], "removed": ["message-1"], "newQueryState": "query-state-2", "hasMoreChanges": false }), "changes"),
    ]).await;
    let result = poll_fastmail(
        request(
            &server.base_url,
            json!({ "queryState": "query-state-1", "mailboxId": "inbox", "sentMailboxId": "sent" }),
        ),
        &http_client().unwrap(),
        "source-only-token",
    )
    .await
    .unwrap();
    assert!(result.items.is_empty());
    assert_eq!(result.checkpoint["queryState"], "query-state-2");
    server.finish().await;
}

#[tokio::test]
async fn enforces_body_thread_message_and_attachment_limits() {
    let mut messages = Vec::new();
    for index in 0..=100 {
        let marker = if index == 0 {
            "UNWANTED_FIRST".to_owned()
        } else {
            format!("body-{index}")
        };
        messages.push(email(
            &format!("message-{index}"),
            "2026-08-03T10:00:00.000Z",
            &format!("{marker}{}", "x".repeat(3_000)),
        ));
    }
    let selected = messages[100].clone();
    let thread_messages = messages[1..].to_vec();
    let ids: Vec<Value> = messages
        .iter()
        .map(|message| message["id"].clone())
        .collect();
    let server = FixtureServer::start(move |base| {
        vec![
            session(base),
            jmap(
                "Email/queryChanges",
                json!({ "added": [{ "id": "message-100", "index": 0 }], "removed": [], "newQueryState": "query-state-2", "hasMoreChanges": false }),
                "changes",
            ),
            jmap("Email/get", json!({ "list": [selected] }), "emails"),
            jmap(
                "Thread/get",
                json!({ "list": [{ "id": "thread-1", "emailIds": ids }] }),
                "thread",
            ),
            jmap(
                "Email/get",
                json!({ "list": thread_messages }),
                "emails",
            ),
        ]
    })
    .await;
    let result = poll_fastmail(
        request(
            &server.base_url,
            json!({ "queryState": "query-state-1", "mailboxId": "inbox", "sentMailboxId": "sent" }),
        ),
        &http_client().unwrap(),
        "source-only-token",
    )
    .await
    .unwrap();
    let item = &result.items[0];
    assert_eq!(item.metadata["threadMessageCount"], 100);
    assert_eq!(item.metadata["attachments"].as_array().unwrap().len(), 100);
    assert_eq!(item.body.encode_utf16().count(), 64 * 1024 * 4);
    assert!(!item.body.contains("UNWANTED_FIRST"));
    server.finish().await;
}

#[tokio::test]
async fn source_binary_writes_one_bounded_response_without_stdout_diagnostics() {
    let server = FixtureServer::start(|base| vec![
        session(base),
        jmap("Mailbox/get", json!({ "list": [{ "id": "inbox", "role": "inbox" }, { "id": "sent", "role": "sent" }] }), "mailboxes"),
        jmap("Email/query", json!({ "queryState": "query-state-1", "ids": [] }), "query"),
    ]).await;
    let mut request = request(&server.base_url, Value::Null);
    request.item_limit = 1;
    let mut child = Command::new(env!("CARGO_BIN_EXE_intake-fastmail-source"))
        .env_clear()
        .env("FASTMAIL_API_TOKEN", "binary-token")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .unwrap();
    child
        .stdin
        .take()
        .unwrap()
        .write_all(&serde_json::to_vec(&request).unwrap())
        .await
        .unwrap();
    let output = child.wait_with_output().await.unwrap();
    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    assert!(output.stderr.is_empty());
    assert_eq!(
        output.stdout.iter().filter(|byte| **byte == b'\n').count(),
        1
    );
    assert!(!String::from_utf8_lossy(&output.stdout).contains("binary-token"));
    let response: PollResponse = serde_json::from_slice(&output.stdout).unwrap();
    assert!(response.items.len() <= request.item_limit);
    let calls = server.finish().await;
    assert!(calls.iter().all(|call| {
        !serde_json::to_string(&call.body)
            .unwrap()
            .contains("binary-token")
    }));
}
